crates/kv-store/src/sstable.rs

```diff
@@ -0,0 +1,1026 @@
+                    XXH_SEED,
+                )
+            {
+                return Err(LoroError::DecodeChecksumMismatchError);
+            }
+        }
+        Ok(())
+    }
+
+    pub fn find_block_idx(&self, key: &[u8]) -> usize {
+        self.meta
+            .partition_point(|meta| meta.first_key <= key)
+            .saturating_sub(1)
+    }
+
+    pub fn find_back_block_idx(&self, key: &[u8]) -> usize {
+        self.meta
+            .partition_point(|meta| meta.last_key.as_ref().unwrap_or(&meta.first_key) <= key)
+            .min(self.meta.len() - 1)
+    }
+
+    fn read_block(&self, block_idx: usize) -> Arc<Block> {
+        let offset = self.meta[block_idx].offset;
+        let offset_end = self
+            .meta
+            .get(block_idx + 1)
+            .map_or(self.meta_offset, |m| m.offset);
+        let raw_block_and_check = self.data.slice(offset..offset_end);
+        Arc::new(Block::decode(
+            raw_block_and_check,
+            self.meta[block_idx].is_large,
+            self.meta[block_idx].first_key.clone(),
+            self.meta[block_idx].compression_type,
+        ))
+    }
+
+    pub(crate) fn read_block_cached(&self, block_idx: usize) -> Arc<Block> {
+        self.block_cache
+            .get_or_insert_with(&block_idx, || {
+                Ok::<_, LoroError>(self.read_block(block_idx))
+            })
+            .unwrap()
+    }
+
+    pub fn contains_key(&self, key: &[u8]) -> bool {
+        if self.first_key > key || self.last_key < key {
+            return false;
+        }
+        let idx = self.find_block_idx(key);
+        let block = self.read_block_cached(idx);
+        let block_iter = BlockIter::new_seek_to_key(block, key);
+        block_iter.next_curr_key() == Some(Bytes::copy_from_slice(key))
+    }
+
+    #[allow(unused)]
+    pub fn get(&self, key: &[u8]) -> Option<Bytes> {
+        if self.first_key > key || self.last_key < key {
+            return None;
+        }
+        let idx = self.find_block_idx(key);
+        let block = self.read_block_cached(idx);
+        let block_iter = BlockIter::new_seek_to_key(block, key);
+        block_iter.next_curr_key().and_then(|k| {
+            if k == key {
+                block_iter.next_curr_value()
+            } else {
+                None
+            }
+        })
+    }
+
+    pub fn valid_keys(&self) -> &FxHashSet<Bytes> {
+        self.keys.get_or_init(|| {
+            let mut keys = FxHashSet::default();
+            for (k, _) in self.iter() {
+                keys.insert(k);
+            }
+            keys
+        })
+    }
+
+    pub fn data_size(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn meta_len(&self) -> usize {
+        self.meta.len()
+    }
+}
+
+#[derive(Clone)]
+pub struct SsTableIter<'a> {
+    table: &'a SsTable,
+    next_block_iter: BlockIter,
+    back_block_iter: BlockIter,
+    next_block_idx: usize,
+    back_block_idx: isize,
+    next_first: bool,
```

Comment:
What does this mean?

Reply:
I should choose a better name. When next or back block iter starts using the same block iter, this parameter will be set to true.